RoddyCore/src/de/dkfz/roddy/tools/ComplexLine.groovy

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2017 eilslabs.
+ *
+ * Distributed under the MIT License (license terms are at https://www.github.com/eilslabs/Roddy/LICENSE.txt).
+ */
+
+package de.dkfz.roddy.tools
+
+import groovy.transform.CompileStatic
+
+/**
+ * A complex line representing e.g. a command line.
+ *
+ * The complex line will try to identify enclosing single or double ticks.
+ *
+ * Created by heinold on 16.01.17.
+ */
+@CompileStatic
+class ComplexLine {
+
+    static Map<Character, Character> possibleContainersByOpeningCharacter = [:]
+    static Map<Character, Character> possibleContainersByClosingCharacter = [:]
+
+    static {
+        possibleContainersByOpeningCharacter['\'' as char] = '\'' as Character
+        possibleContainersByOpeningCharacter['\"' as char] = '\"' as Character
+        possibleContainersByOpeningCharacter['{' as char] = '}' as Character
+        possibleContainersByOpeningCharacter['[' as char] = ']' as Character
+        possibleContainersByOpeningCharacter['(' as char] = ')' as Character
+
+        possibleContainersByOpeningCharacter.each { Character key, Character value -> possibleContainersByClosingCharacter[value] = key }
+    }
+
+    String line;
+    LineNode parsedLine;
+
+    /**
+     * Parsed line represents a transformed / parsed line. Currently, single and double ticks will be used
+     * to identify children.
+     *
+     * E.g. --abc --def="ab,def;efk=test;anotherValue='equals'" --jki --a="something=def"
+     *
+     * will be represented like:
+     *
+     *     --abc --def="###CHILD_0###" --jki --a="###CHILD_1###"
+     *                       |                         |
+     *                       ab,def;efk=test;anotherValue="###CHILD_0###"
+     *                                                 |        |
+     *                                                 something=def
+     *                                                          |
+     *                                                          equals
+     */
+    static abstract class LineNode {
+
+        abstract String reassemble()
+
+        abstract String[] splitBy(String s)
+
+        @Override
+        public String toString() {
+            return reassemble()
+        }
+    }
+
+    static class TextNode extends LineNode {
+        StringBuilder text = new StringBuilder()
+
+        TextNode() {
+
+        }
+
+        TextNode(String start) {
+            this.text << start
+        }
+
+        @Override
+        String reassemble() {
+            return text.toString()
+        }
+
+        @Override
+        String[] splitBy(String s) {
+            return text.toString().split(s)
```

Comment:
Maybe this would be a good place to handle special situations, such as ",blabla" or "blabla," when splitting on ','. Maybe you wouldn't need the extra + "" below then.